redhat-cop/agnosticv
# Add support for custom merge strategies

This change, if applied, adds support for merge-strategy customizations in agnosticV.

With this change, it's now possible to configure how agnosticv will
merge the variables, by specifying a list of `[path, strategy]` directly
inside a schema.

for example:

```yaml
type: object
x-merge:
  - path: /__meta__/access_control
    strategy: overwrite
properties:
  [...]
```

Also:

- Get rid of YAML merge, and use JSON-compatible merge only. It's too
  cumbersome to work with both kind of map: `map[any]any` (yaml)
  `map[string]any` (json-compatible).  It complicates the
  implementation too much to support both types everywhere. Plus we
  should never use something else that string for keys of dictionary.
- reorganize code in sperate files per topic (merge, schema)
- Update to go1.18 and replace interface{} with any
- Test overwrite on a `/__meta__/access`
- remove unecessary merge in main program now that we only use
  "jsonyaml" merge
- Create a new schema2 to test multischema scenarii
- Add unit test to ensure overwite is working for `/__meta__/access_control`
- Update readme

This code has been tested against our production agnosticV catalog and previous version `v0.3.3`, to produce exactly the same list and merged vars:

```
~/sync/dev/agnosticv-public/tools/test_new_version.sh ~/sync/dev/agnosticv ~/bin/agnosticv.v0.3.3 ~/bin/agnosticv.GPTEINFRA-3014
```
